### PR TITLE
Feature/middleware

### PR DESCRIPTION
**Pull request by haf**
This implements middlware for logary and brings the C# API back in line with v4. It also adds the `time` function to the `Logging` module.

### CODE REVIEW

File: src/Logary.CSharp/LoggerExtensions.cs [hunk 1]
```diff
@@ -106,203 +108,41 @@ public static T Time<T>(this Logger logger, Func<T> f, LogLevel level = null)
...
-            level = level ?? LogLevel.Debug;
-            a();
+            Time (logger, () => {
+                a ();
+                return 0;
+            }, level);
         }
 
         /// <summary>
         /// TBD
```
Comment (neoeinstein) [inline src/Logary.CSharp/LoggerExtensions.cs]: Update the XML doc now that the implementation is provided.

File: src/Logary/Configuration_Config.fs [hunk 2]
```diff
@@ -104,6 +105,14 @@ let withMetrics ms conf =
   |> Seq.map Metric.validate
   |> Seq.fold (fun s t -> s |> withMetric t) conf
 
+[<CompiledName "WithMiddleware">]
+let withMiddleware middleware conf =
+  { conf with LogaryConf.middleware = middleware :: conf.middleware }
```
Comment (neoeinstein) [inline src/Logary/Configuration_Config.fs]: Does/will ordering of middlewares matter? This looks to be consistent with your concatenation of multiple middlewares below, but we will want to be explicit about how the ordering affects the application.

Comment (haf) [inline src/Logary/Configuration_Config.fs]: Middleware registered first will run first. So

``` fsharp
|> withMiddleware messageId
|> withMiddleware principalDetails
```

will first run messageId, then `principalDetails`. So if `messageId` is the same middleware as we have in logary-js, then this order would be in error.

**Comment by futurechan** [comment]
👍

**Pull request merged**